File: Week_03/98/validate_binary_search_tree.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


class Solution:
    def isValidBST(self, root: TreeNode) -> bool:
        return self.dfs(root, -2 ** 63, 2 ** 64 - 1)

    def dfs(self, node: TreeNode, left: int, right: int) -> bool:
        if node is None:
            return True

        return left < node.val < right and self.dfs(node.left, left, node.val) and self.dfs(node.right, node.val, right)

File: Week_03/98/validate_binary_search_tree.go
package main

import "math"

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func dfs(node *TreeNode, left, right int) bool {
	if node == nil {
		return true
	}
	return node.Val > left && node.Val < right &&
		dfs(node.Left, left, node.Val) && dfs(node.Right, node.Val, right)
}
func isValidBST(root *TreeNode) bool {
	return dfs(root, math.MinInt64, math.MaxInt64)
}
